### DISCUSSION TITLE
[Idea] $accordion - your input please

### DISCUSSION THREAD

**Discussion by twMat** [TiddlyWiki/TiddlyWiki5#5839] [Ideas]
[Would like input on this proposal before posting as issue. Thanx.]

[edit: Idea submitted as Issue request #6185 ]

"**Accordions**" a.k.a "**sliders**" are popular and, for an end user, on par with, say, a checkbox or selectwidget. But, an accordion is not a widget. Instead it has to be created like so:

```
<$reveal type="nomatch" state="$:/state/SampleReveal2" text="show">

<$button set="$:/state/SampleReveal2" setTo="show">Show me</$button>

</$reveal>
<$reveal type="match" state="$:/state/SampleReveal2" text="show">

<$button set="$:/state/SampleReveal2" setTo="hide">Hide me</$button>

! This is the revealed content
And this is some text

</$reveal>
```

This is *very* complicated: It involves (and nests) multiple widgets, there's the clever-but-confusing `match/nomatch - show/hide` interplay, and there's the manual handling of the `state` tiddler. IMO these complexities don't benefit the user - they are instead there *only* to achieve the construct itself. For example, for the `state` tiddler the user has to type it in four places (after coming up with the name to begin with) but this state is thereafter typically never used for anything *else*!

So, I'd like a designated, user friendly, and substantially simpler, "accordion widget". Here's an idea: 

```
<$accordion labelClosed="show" labelOpen="hide" type="popup" class="myClass">

! This is the revealed content
And this is some text

</$accordion>
```

I'd think most of it is self explanatory. All parameters are optional but notably:

* `labelClosed` is synonymous to `label`. If this parameter is omitted or if no argument is set, the label defaults to e.g  `≫` or `⊕`.
* Similarly, if `labelOpen` (`labelOpened`?) is declared but without an argument then e.g `⩔` is used.

IMO, a most important matter is that **state is handled internally**. I'm assuming the state can be named via a randomized variable - yes?

One might argue that this should be a *macro* but the macro syntax format requires that "content" resides between the `<<` and `>>` which is not good for free wikitext content.

Maybe names like `$fold` or `$expand` would be more typical widget names, but I felt the light-heartedness of `$accordion` signals a friendliness that I expect beginners to appreciate.

**Comment by AnthonyMuscio** [comment]
Mat,

I applaud your suggestion. Yet I believe the answer is understanding how to use the html `<details>` tag or the [$details plugin widget](https://tid.li/tw5/plugins.html#DetailsWidget). I have being playing with this and trying to make an simple, easy to use solution. Perhaps what you left out of the $accordion requirements includes is a initial state, class/style options, block/inline options and more. But this is the right place to develop this idea further.

With my own exploration of this there is a whole suite of solutions that do not need a state tiddler, the is the desire for a simple "more..." and also nested accordion's. I often remember the old nested sliders plugin in TWC using the +++

Of course it is also possible to transclude you content or define it in a macro and use a variable to access it. This can be helpful in Question answer pairs because the answer is hidden.

Regards
Tony

**Reply by AnthonyMuscio** [reply]
Post script, 

- if the solution also allowed open and close actions all the better. A work around for now is to use the checkbox or RadioWidget as the trigger to open close a section of code.
- CSS is also a contender, eg make your own html tag `<accordion class=accordion style="">content</accordion>` and use of the data tags with css. The ID attribute can help targeted accordions.

**Reply by twMat** [reply]
@AnthonyMuscio I had indeed forgotten about the $details plugin! Still, this feature is too basic and popular to not have natively. Also, the details plugin is very opinionated but I guess that could be trimmed.

But then, when people request features for the standard distro, the typical dismissal is to first try it out as a plugin and if it proves popular and useful then it can be considered for core... I'd say the details plugin is very much such a case..

**Comment by AnthonyMuscio** [comment]
As I look at this further there are a range of uses for an according like tool, and example may be an accordion for answers to questions, a simple more button and others.  Each of which could benefit from custom methods for each type of accordion. 

- I think Shiraz provides methods that would be inspirational.
- I am researching a method to use "buttons" to alter css to give us the state change not unlike the details tag and widget. 
- We can open a class of accordions in one hit or a unique one using the ID attribute.

I have confirmed the CSS based solution is effective. Watch this space.

**Reply by twMat** [reply]
Let's not mix in applications at this stage even if, of course, one has to keep them in mind to make sure the product doesn't prevent them. But, as stated previously, what I'm after is a *simple* reveal type widget. If one wants power, then the current button+reveal can solve it.

>I have confirmed the CSS based solution is effective. Watch this space.

Sure! :-)

**Reply by AnthonyMuscio** [reply]
I now have a working solution, I have designed it to handle more, Questions and their answers and a few more;

Basically you create a section using arbitrary html tags, give each an id
```
<more id=m1>
all your desired content here including transclusions and wiki text.
</more>
```

Then provide the "toggle" eg; before or after the above more content, via a macro giving the id
`<<more m1>>`

- **Is this above method clean and easy?**
- Is it effectively a "is a simple reveal"

I may be able to hide the need to use the macro (not sure yet)

**How it works**
The status of the more is stored in a shared datatiddler with the ID as the key, its status as the value.

- A stylesheet determines the default for for any of these arbitrary html tags in use like `<more> `
- typically this will be `display: none;`
- When you click the more button which looks like "more...", another button is displayed "less..." along with; some css in form;
`<style>more id=N { display: block }</style>` which causes the more content to be displayed.

Since the above uses ID's they need to be unique, but something similar can be done with classes. 

Also much of this can be simplified further and automated with Mario's Custom markup project.
eg; the wiki text would look something like this.

```
°more::m1
content 
/°
```

Or with optional class(es) to colorise, indent etc...
```
°more.class1::m1
content 
/°
```

Feedback sought!

Tones

**Reply by twMat** [reply]
@AnthonyMuscio - sounds interesting. Could you put up a tiddlyhost demo?

@ everyone  
....but, regardless, I'd assume a non-wikitext solution is appropriate for the standard distro (why actually!?) so my OP question still holds. Why is the proposed $accordion widget a good/bad idea? Or can non-wikitext things like Tones thing be wrapped into widgets?

**Reply by AnthonyMuscio** [reply]
Further to this thread, I have worked out a tiddlywiki way to set unique html/css ID's which can be addressed eg: to force display: block; or display: none; to toggle visibility.

Here is some test code;

I believe I can soon find a way to use css counters to automatically count a particular tag used in a tiddler.

```
<style>
question {
  counter-increment: qnumber;
  display: block;
}

question::before {
  content:counter(qnumber);
}
question::after {
  content: " (" attr(id) ")";
}
</style>

<div>
<question id={{{ [{!!title}addprefix[q1-]slugify[]] }}}>
one q
</question>
<question id={{{ [{!!title}addprefix[q2-]slugify[]] }}}>
two q
</question>
<question id={{{ [{!!title}addprefix[q3-]slugify[]] }}}>
three q
</question>
</div>
```

Regards
Tones

**Comment by Alzacon** [comment]
We have [toc-expandable](https://tiddlywiki.com/#Example%20Table%20of%20Contents%3A%20Expandable) macro which looks "similar" to your idea (IMO). I think that it is possible to start by modifying this toc macro or reuse some parts of its code. Then you can play to show the title of tiddler and display/hide their field text.

**Reply by twMat** [reply]
> toc-expandable

Not sure about this, for several reasons: 1) You have the problem with macros that I mentioned i.e that any arguments to a macro must be added within the `<<...>>`. 2) The toc macros are presumably based around some recursive mechanism so it feels like a less optimal starting point.  3) It contains revealwidgets which demands statetiddlers, which is fine... until you put multiple instances in the same tiddler. This particular problem has to somehow be solved or circumvented in my request.

**Reply by Alzacon** [reply]
We have macrocall widget which solves the problem with variables/macros as parameters.

I don't talk about of take all code, you can reuse few parts. You don't need know if a tiddler is used as tag. You can use the part of code for display but you don't need recurive mechanism, a list widget could be enough.

I think that you can have diffent statetiddllers for intances.

I don't think that the solution is a new widget similar to reveal but make it more versatile.

**Reply by twMat** [reply]
I'm afraid your answers don't make sense to me and the problems I outlined still hold.

**Comment by AnthonyMuscio** [comment]
Mat,

I have found another way to handle this. It makes use of the CSS global display attribute.

I have created a demo especially for you Mat

https://anthonymuscio.github.io/more-or-less

Regards
Tones

**Reply by AnthonyMuscio** [reply]
Some working sample code to tempt you (works on my linked site)

```
<<display block about  "about this wiki..." "...hide about">><about class=none style={{!!about}}>

This Wikis was built to demonstrate an alternative to the idea for an according widget raised by Mat [[here |https://github.com/Jermolene/TiddlyWiki5/discussions/5839]]

<<display block more>><more class=none style={{!!more}}>

Especially for more details... that can be nested
</more>

</about>
```

Tones

**Reply by AnthonyMuscio** [reply]
I have yet another alternative making use of a new feature in 5.2.0 the ability to pass parameters in macros named in filters. 

In this case a simple form of the list widget is used but all the power of the list widget is still available.

There may however be a bug in the new feature I am troubleshooting.

Tony

**Reply by twMat** [reply]
@AnthonyMuscio - aw, man, I have missed this post/solution of yours for me, my apologies. I will look into your solution properly tonight or tomorrow. Much appreciated.

**Reply by twMat** [reply]
OK, I have now checked it out. First of all; thank you very much Tony!

So from what I understand this is the method: There is a macro showing a button to toggle a local field. The local field gets values `display:none` or `display:block`. Outside of the macro call there is also a separate html tag with a `style` attribute transcluding that field value. - Cool idea!

Frankly, I feel it is too complex for *spontaneous* use though, which is what I'm after. Here´s a somewhat simplified version (macro follows below) but it is also not *simple* enough IMO:

```
<<+++ mylabel>>
<mylabel>
My text
</mylabel>
```

The macro:

```
\define +++(label)
\whitespace trim
<$list filter="""[all[current]!has[$label$]]""">
   <$button  class="tc-btn-invisible bluelink">$label$
      <$action-setfield $field="$label$" $value="active"/>
   </$button>
   <style>
      [data-tiddler-title=<<currentTiddler>>] $label$ {display:none;}
   </style>
</$list>
<$list filter="""[all[current]has[$label$]]""">
   <$button class="tc-btn-invisible bluelink">$label$
      <$action-deletefield $label$/>
   </$button>
   <style>
     [data-tiddler-title=<<currentTiddler>>] $label$ {display:block;}
   </style>
</$list>
\end

<style>.bluelink {color:#5778d8;}</style>
```

As you can see, a big limitation is that the label has to stay within "style definition name format" (no spaces, no Uppercase, ...). This can probably be solved, but, regardless, I still feel it is still not *convenient* enough when typing freely. A construct like the above is more suitable when one is "constructing" rather than "typing"

To clarify: "Folding" is so *useful* that we should compare it with typing bullet lists or tables in wikitext - and these are as simple as `* bullet list` and `|table|construct|`. Just imagine if we could just type e.g ` >this And we got a foldable segment< ` (where `this` is a label of some sort)

But, OK, that is a fantasy so back to our construction: I wish we could some up with something that at least only needs  *two* "parts" to enclose the text - i.e one *prefixing* thingy and one *suffixing* thingy - instead of the above three "parts" (the macrocall+two tags). If one wants the button to be *distanced* from the revealed text then, sure, a separate button is good(!) But when *typing*, the typical scenario is that one wants it where one is at.

**Comment by AnthonyMuscio** [comment]
See also David's [nested details discussion](https://groups.google.com/g/tiddlywiki/c/gCp5ONBxFXg)

**Comment by twMat** [comment]
As I say in the OP, this thread is to get input for an issue request so everyones opinions are appreciated.

The last post of mine made me realize that what would really be useful for *freely* being able to use folding is the most simplified *wikitext markup* form, comparable to 

```
* wikitext 
* bullet
* lists
```
and

```
|wiki|text|
|tables||
```

...now *that* is simple and it minimally disturbs the typing. IMO "folding" is also as *useful* as bullet list or tables, or it would be if it was as simple to use. **So, what is a possible markup for this**?
Just as an example, here's the syntax I improvised in the previous post:

` >label My text< ` - i.e "space + greaterthan" directly followed by a no-space label+space then the content until the closing smallerthan. Just like with e.g a `|table|` you can't just put anything in it, that's fair. And could that syntax even be nestable? 

What is a better syntax? Is the idea at all implementable?

**Reply by Alzacon** [reply]
Is wikitext for Details (element)? It sounds interesting.

I think that `space + &gt` has conflict with the quote block.

I think that we could use wikirules similar to the parts of Description List element ( `:` and `;` ). The problem is in conflicts with other wikirules (and solve it) or use markup that any user would type. I think in `->` and `-<` but in the first is a "arrow" that users can maybe used it or there are conflict with some plugin. I don't know if it is implementable in the core. For example, Mario have a plugin for custom markup and other for comments (and if I remember correctly the wiki rule for comments was proposed)

**Comment by twMat** [comment]
Idea submitted as Issue request #6185
